[FLINK-5229] [state] Cleanup of operator snapshots if subsequent operator snapshots fail

tillrohrmann opened the pull request:
This PR is based on #3178.

This PR adds operator state cleanup to the StreamTask class. If a stream task contains multiple
stream operators, then every operator is checkpointed. In case that a snapshot operation fails
all state handles and OperatorSnapshotResults belonging to previous operators have to be freed.

StefanRRichter commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -954,6 +954,27 @@ public void run() {
+					}
+				}
+
+				for (StreamStateHandle nonPartitionedStateHandle : nonPartitionedStateHandles) {
```
You could use the utility method `StateUtil.bestEffortDiscardAllStateObjects(...)` to reduce code duplication a bit.

I think we should also discard other state objects that completed fuctures could have created, e.g. `StateUtil.bestEffortDiscardAllStateObjects(operatorStatesBackend);` etc. , what do you think?

tillrohrmann replied:
Yes you're absolutely right. Will address your comments.

tillrohrmann replied:
Having a second look, I think the other state objects don't have to be cleaned up, because they are still owned by the corresponding `OperatorSnapshotResult` which is cleaned up in case of a failure or `close` operation.

StefanRRichter replied:
Agreed :-)

tillrohrmann commented:
Thanks for the review @StefanRRichter. Will address your comments.

tillrohrmann commented:
I've addressed your comments @StefanRRichter. Note that I did not add the additional state object cleanup, because it was already done by cancelling the `OperatorSnapshotResults`.
